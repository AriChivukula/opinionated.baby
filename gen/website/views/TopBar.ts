import { Bespoke, Import, Module, Renderable } from "typescriptase";

export const module: Renderable = Module.new({
  content: [
    Bespoke.new({name: "DEPRECATE"}),
  ],
  destination: "src/website/views/TopBar.tsx",
  imports: [
    Import.new({
      module: "js-cookie",
      nameAll: "cookie",
    }),
    Import.new({
      module: "react",
      nameAll: "React",
    }),
    Import.new({
      module: "react-relay",
      names: ["commitMutation", "createFragmentContainer", "graphql"],
    }),
    Import.new({
      module: "relay-runtime",
      nameDefault: "Relay",
    }),
    Import.new({
      module: "rmwc",
      names: ["Toolbar", "ToolbarFixedAdjust", "ToolbarIcon", "ToolbarRow", "ToolbarSection", "ToolbarTitle"],
    }),
    Import.new({
      module: "url",
      names: ["parse", "UrlWithParsedQuery"],
    }),
    Import.new({
      module: "../util",
      names: ["goto"],
    }),
    Import.new({
      module: "./__generated__/TopBarLoginMutation.graphql",
      names: ["TopBarLoginMutationResponse"],
    }),
    Import.new({
      module: "./__generated__/TopBarLogoutMutation.graphql",
      names: ["TopBarLogoutMutationResponse"],
    }),
    Import.new({
      module: "./__generated__/TopBarQuery.graphql",
      names: ["TopBarQuery"],
    }),
  ],
});
