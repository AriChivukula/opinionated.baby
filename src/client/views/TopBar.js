// @flow

import type {TopBarQuery} from './__generated__/TopBarQuery.graphql.js';

import cookie from 'js-cookie';
import React, { Component } from 'react';
import {
  createFragmentContainer,
  commitMutation,
  graphql
} from 'react-relay';
import {
  Environment,
} from 'relay-runtime';
import {
  Card,
  CardPrimaryAction,
  CardMedia,
  CardAction,
  CardActions,
  CardActionButtons,
  CardActionIcons
} from 'rmwc/Card';
import {
  Grid,
  GridCell,
  GridInner
} from 'rmwc/Grid';
import {
  List,
  SimpleListItem
} from 'rmwc/List';
import {
  Toolbar,
  ToolbarFixedAdjust,
  ToolbarIcon,
  ToolbarRow,
  ToolbarSection,
  ToolbarTitle,
} from 'rmwc/Toolbar';
import { Typography } from 'rmwc/Typography';
import url from 'url';

type Props = {
  data: ?TopBarQuery,
  blockUntilReload: () => void,
  relay: {
    environment: Environment
  }
}

class TopBar extends Component<Props> {

  static defaultProps = {
    relay: {
      environment: null
    }
  }

  componentDidMount() {
    const url_parts = url.parse(window.location.href, true);
    if (url_parts.query && url_parts.query.code) {
      this.login(url_parts.query.code);
    }
  }

  render() {
    let login = (
      <ToolbarIcon
        use="person"
        onClick={() => this.googleAuth()}
      />
    );
    const data = this.props.data;
    if (data != null) {
      const me = data.me;
      if (me != null) {
        login = <React.Fragment>
          <ToolbarTitle>{me.email}</ToolbarTitle>
          <ToolbarIcon
            use="exit_to_app"
            onClick={() => this.logout()}
          />
        </React.Fragment>;
      }
    }
    return (
      <div>
        <Toolbar fixed>
          <ToolbarRow>
            <ToolbarSection alignStart>
              <ToolbarTitle>Opinionated Baby</ToolbarTitle>
            </ToolbarSection>
            <ToolbarSection alignEnd>
              {login}
              <ToolbarIcon
                use="code"
                onClick={() => this.goto('https://github.com/arichiv/opinionated.baby/')}
              />
              <ToolbarIcon
                use="info"
                onClick={() => this.goto('http://chivuku.la/')}
              />
            </ToolbarSection>
          </ToolbarRow>
        </Toolbar>
        <ToolbarFixedAdjust />
        <Grid>
          <GridCell span={12}>
            <Typography use="display2" tag="div">
              A confidently immature starting point
            </Typography>
          </GridCell>
          <GridCell span={12}>
            <Typography use="subheading2" tag="div">
              The hardest part of building a new dynamic websites is avoiding
              all the dragons. Starting a development environment can freeze us
              with choice overload.
              {' '}
              <Typography theme="text-secondary-on-background">
                Opinionated Baby
              </Typography>
              {' '}
              solves this by moving the starting line past the quagmire of setup
              and into the pleasure of production logic.
            </Typography>
          </GridCell>
          <GridCell span={12}>
            <Typography use="display2" tag="div">
              Important choices I made for you
            </Typography>
          </GridCell>
          <GridCell span={3}>
            <List twoLine>
              <SimpleListItem
                graphic="memory"
                text="Lambda"
                secondaryText="Server(less)"
                onClick={() => this.goto('https://aws.amazon.com/lambda/')}
              />
              <SimpleListItem
                graphic="share"
                text="PostgresSQL"
                secondaryText="Database"
                onClick={() => this.goto('https://www.postgresql.org/')}
              />
              <SimpleListItem
                graphic="storage"
                text="S3"
                secondaryText="Storage"
                onClick={() => this.goto('https://aws.amazon.com/s3/')}
              />
              <SimpleListItem
                graphic="bug_report"
                text="Travis CI"
                secondaryText="CI/CD"
                onClick={() => this.goto('https://travis-ci.org/')}
              />
            </List>
          </GridCell>
          <GridCell span={3}>
            <List twoLine>
              <SimpleListItem
                graphic="description"
                text="JavaScript"
                secondaryText="Language"
                onClick={() => this.goto('https://developer.mozilla.org/en-US/docs/Web/JavaScript/')}
              />
              <SimpleListItem
                graphic="language"
                text="Babel"
                secondaryText="Transpiler"
                onClick={() => this.goto('https://babeljs.io/')}
              />
              <SimpleListItem
                graphic="merge_type"
                text="Flow"
                secondaryText="Typing"
                onClick={() => this.goto('https://flow.org/')}
              />
              <SimpleListItem
                graphic="cached"
                text="Node.js"
                secondaryText="Runtime"
                onClick={() => this.goto('https://nodejs.org/en/')}
              />
            </List>
          </GridCell>
          <GridCell span={3}>
            <List twoLine>
              <SimpleListItem
                graphic="unarchive"
                text="yarn"
                secondaryText="Dependency"
                onClick={() => this.goto('https://yarnpkg.com/en/')}
              />
              <SimpleListItem
                graphic="build"
                text="Gulp"
                secondaryText="Toolchain"
                onClick={() => this.goto('https://gulpjs.com/')}
              />
              <SimpleListItem
                graphic="playlist_add_check"
                text="Jest"
                secondaryText="Testing"
                onClick={() => this.goto('https://facebook.github.io/jest/')}
              />
              <SimpleListItem
                graphic="graphic_eq"
                text="GraphQL"
                secondaryText="API"
                onClick={() => this.goto('https://graphql.org/')}
              />
            </List>
          </GridCell>
          <GridCell span={3}>
            <List twoLine>
              <SimpleListItem
                graphic="flip_to_front"
                text="React"
                secondaryText="Interface"
                onClick={() => this.goto('https://reactjs.org/')}
              />
              <SimpleListItem
                graphic="local_library"
                text="Relay"
                secondaryText="Fetch"
                onClick={() => this.goto('http://facebook.github.io/relay/')}
              />
              <SimpleListItem
                graphic="style"
                text="Material Design"
                secondaryText="Theme"
                onClick={() => this.goto('https://material.io/components/web/')}
              />
              <SimpleListItem
                graphic="line_style"
                text="SASS"
                secondaryText="Style"
                onClick={() => this.goto('https://sass-lang.com/')}
              />
            </List>
          </GridCell>
          <GridCell span={12}>
            <Typography use="display2" tag="div">
              <span style={{textDecoration: 'line-through'}}>
                Self justification
              </span>
              {' '}
              Philosophy
            </Typography>
          </GridCell>
          <GridCell span={12}>
            <Typography use="subheading2" tag="div">
              This framework is one-size-fits-me. It{'\''}s an open source
              version of a system I use when building complex websites.
              Anyone may attempt to use, contribute, or critique it. The
              consistency with which feedback is incorporated will be nothing
              short of capricious. My general mission and purpose is to support
              the vitality and happy, healthy development of our
              {' '}
              <a href={'http://www.nic.baby/policies.html'}>
                babies and children
              </a>
              .
            </Typography>
          </GridCell>
          <GridCell span={12}>
            <Typography use="display2" tag="div">
              Release history
            </Typography>
          </GridCell>
          <GridCell span={4}>
            <Card>
              <CardPrimaryAction onClick={() => this.goto('https://github.com/arichiv/opinionated.baby/releases/tag/v0/')}>
                <CardMedia
                  square
                  style={{backgroundImage: 'url(images/v0.jpg)'}}
                />
                <Typography
                  use="display1"
                  style={{padding: '1rem'}}>
                  <Typography theme="text-secondary-on-background">
                    v0
                  </Typography>
                  {' '}
                  Cocksure Castle
                </Typography>
              </CardPrimaryAction>
            </Card>
          </GridCell>
        </Grid>
      </div>
    );
  }

  googleAuth(): void {
    const data = this.props.data;
    if (data != null) {
      window.location = data.loginURL;
    }
  }

  login(code: string): void {
    commitMutation(
      this.props.relay.environment,
      {
        mutation: graphql`
          mutation TopBarLoginMutation($code: String!) {
            login(code: $code) {
              accessToken
            }
          }
        `,
        variables: {
          code: code
        },
        onCompleted: (response, errors) => {
          cookie.set('access_token', response.login.accessToken);
          window.location = '/';
        }
      }
    );
    this.props.blockUntilReload();
  }

  logout(): void {
    commitMutation(
      this.props.relay.environment,
      {
        mutation: graphql`
          mutation TopBarLogoutMutation {
            logout {
              accessToken
            }
          }
        `,
        variables: {},
        onCompleted: (response, errors) => {
          cookie.set('access_token', response.logout.accessToken);
          window.location = '/';
        }
      }
    );
    this.props.blockUntilReload();
  }

  goto(url: string): void {
    window.location = url;
  }
}

export default createFragmentContainer(
  TopBar,
  graphql`
    fragment TopBarQuery on Query @argumentDefinitions(
      access_token: {type: "String!"}
    ) {
      loginURL,
      me(access_token: $access_token) {
        googleID,
        email
      }
    }
  `,
);
