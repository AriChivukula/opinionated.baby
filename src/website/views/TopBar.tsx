/**
 * This file is partially generated; only edit bespoke sections.
 *
 * SOURCE<<gen/website/views.ts::TopBarReact>>
 * BESPOKE<<imports, render, implementation, relay>>
 * SIGNED<<cc1AN1qU5ZOxiGxyc9wF1QlFpobHngrvXZxS4gb4MbFdF9Kwsq6BXfxlyscRAxaEjJCZjKPpCBzJFES+sPkP3g==>>
 */

import * as React from "react";
import {
  polyfill,
} from "react-lifecycles-compat";
import {
  commitMutation,
  createFragmentContainer,
  graphql,
} from "react-relay";

/* BESPOKE START <<imports>> */
import * as cookie from "js-cookie";

import Relay from "relay-runtime";
import {
  Toolbar,
  ToolbarFixedAdjust,
  ToolbarIcon,
  ToolbarRow,
  ToolbarSection,
  ToolbarTitle,
} from "rmwc/Toolbar";
import {
  Url,
} from "url";

import {
  goto,
} from "../util";
import {
  TopBarLoginMutationResponse,
} from "./__generated__/TopBarLoginMutation.graphql";
import {
  TopBarLogoutMutationResponse,
} from "./__generated__/TopBarLogoutMutation.graphql";
import {
  TopBarQuery,
} from "./__generated__/TopBarQuery.graphql";
/* BESPOKE END <<imports>> */

export interface ITopBarProps {
  data: TopBarQuery;
  relay: Relay;
}

class __TopBar extends React.Component<ITopBarProps> {

  public render(
  ): JSX.Element {
    /* BESPOKE START <<render>> */
    const urlParts: URL = new URL(window.location.href);
    if (urlParts.searchParams.has("code")) {
      this.login(urlParts.searchParams.get("code") as string);

      return <div />;
    }
    let login: JSX.Element = (
      <ToolbarIcon
        use="person"
        onClick={(): void => { this.googleAuth(); }}
      />
    );
    if (this.props.data.me !== null) {
      login = (
        <>
          <ToolbarTitle>{this.props.data.me.email}</ToolbarTitle>
          <ToolbarIcon
            use="exit_to_app"
            onClick={(): void => { this.logout(); }}
          />
        </>
      );
    }

    return (
      <>
        <Toolbar fixed waterfall>
          <ToolbarRow>
            <ToolbarSection alignStart>
              <ToolbarTitle onClick={(): void => { goto("https://github.com/arichiv/opinionated.baby/"); }}>
                {document.title}
              </ToolbarTitle>
            </ToolbarSection>
            <ToolbarSection alignEnd>
              {login}
              <ToolbarIcon
                use="code"
                onClick={(): void => { goto("https://github.com/arichiv/opinionated.baby/"); }}
              />
              <ToolbarIcon
                use="info"
                onClick={(): void => { goto("http://chivuku.la/"); }}
              />
            </ToolbarSection>
          </ToolbarRow>
        </Toolbar>
        <ToolbarFixedAdjust />
      </>
    );
    /* BESPOKE END <<render>> */
  }

  /* BESPOKE START <<implementation>> */
  private googleAuth(
  ): void {
    if ("loginURL" in this.props.data) {
      goto(this.props.data.loginURL, true);
    }
  }

  private login(
    code: string,
  ): void {
    commitMutation(
      this.props.relay.environment,
      {
        mutation: graphql`
          mutation TopBarLoginMutation($input: LoginInput) {
            login(input: $input) {
              accessToken
            }
          }
        `,
        onCompleted: (response: TopBarLoginMutationResponse, errors: Error[]): void => {
          cookie.set(
            "accessToken",
            response.login.accessToken,
          );
          goto("/", true);
        },
        variables: {
          input: {
            code,
          },
        },
      },
    );
  }

  private logout(
  ): void {
    commitMutation(
      this.props.relay.environment,
      {
        mutation: graphql`
          mutation TopBarLogoutMutation($input: LogoutInput) {
            logout(input: $input) {
              accessToken
            }
          }
        `,
        onCompleted: (response: TopBarLogoutMutationResponse, errors: Error[]): void => {
          cookie.set(
            "accessToken",
            response.logout.accessToken,
          );
          goto("/", true);
        },
        variables: {
          input: {
            dummy: "",
          },
        },
      },
    );
  }
  /* BESPOKE END <<implementation>> */
}

const _TopBar: React.ComponentType = createFragmentContainer(
  __TopBar,
  /* BESPOKE START <<relay>> */
  graphql`
    fragment TopBarQuery on Query {
      loginURL,
      me {
        id,
        email
      }
    }
  `,
  /* BESPOKE END <<relay>> */
);

// @ts-lint
export const TopBar: React.Component<ITopBarProps> = polyfill(_TopBar);
