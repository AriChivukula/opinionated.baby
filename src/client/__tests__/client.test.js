// @flow

import 'babel-polyfill';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import context from 'react-test-context-provider';
import {
  Environment,
  RecordSource,
  Store,
} from 'relay-runtime';
import {
  urlMiddleware,
  RelayNetworkLayer
} from 'react-relay-network-modern';

import TopBar from '../views/TopBar.js';

Enzyme.configure({
  adapter: new Adapter()
});

const environment = new Environment({
  network: new RelayNetworkLayer([
    urlMiddleware({
      url: 'http://127.0.0.1:8080',
    })
  ]),
  store: new Store(new RecordSource())
});

test(
  'TopBar',
  async () => {
    const component = Enzyme.render(
      Enzyme.mount(
        context(
          {
            relay:{
              environment: environment,
              variables: {}
            }
          },
          // $FlowFixMe
          <TopBar
            data={{
              __id: '0',
              __fragments: {
                'TopBarQuery': {}
              },
              me: {
                id: 'TEST',
                googleID: 'TEST',
                email: 'TEST'
              },
              loginURL: 'http://fake.com/'
            }}
            blockUntilReload={() => {}}
          />
        )
      )
    );
    expect(component.html()).toMatchSnapshot();
  }
);
