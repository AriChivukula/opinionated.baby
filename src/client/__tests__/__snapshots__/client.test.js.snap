// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": " mdc-toolbar",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": " mdc-toolbar__row",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": " mdc-toolbar__section mdc-toolbar__section--align-start",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": " mdc-toolbar__title",
                },
                "children": Array [
                  Object {
                    "data": "Opinionated Baby",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "section",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": " mdc-toolbar__section mdc-toolbar__section--align-end",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": " mdc-toolbar__title",
                  },
                  "children": Array [
                    Object {
                      "data": "TEST",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "material-icons  mdc-toolbar__icon ",
                    },
                    "children": Array [
                      Object {
                        "data": "exit_to_app",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "i",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                Object {
                  "attribs": Object {
                    "class": "material-icons  mdc-toolbar__icon ",
                  },
                  "children": Array [
                    Object {
                      "data": "exit_to_app",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "i",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": " mdc-toolbar__title",
                    },
                    "children": Array [
                      Object {
                        "data": "TEST",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "section",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": " mdc-toolbar__section mdc-toolbar__section--align-end",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": " mdc-toolbar__title",
                },
                "children": Array [
                  Object {
                    "data": "TEST",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "material-icons  mdc-toolbar__icon ",
                  },
                  "children": Array [
                    Object {
                      "data": "exit_to_app",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "i",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "class": "material-icons  mdc-toolbar__icon ",
                },
                "children": Array [
                  Object {
                    "data": "exit_to_app",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "i",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": " mdc-toolbar__title",
                  },
                  "children": Array [
                    Object {
                      "data": "TEST",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "section",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": " mdc-toolbar__section mdc-toolbar__section--align-start",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": " mdc-toolbar__title",
                  },
                  "children": Array [
                    Object {
                      "data": "Opinionated Baby",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "section",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "header",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
